File: services/client/src/internal/transport/socket7e.go
package transport

import (
	"net"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

type Socket7E struct {
	socket net.Conn
}

var MAGIC_NUMBER = []byte{0x7E}

func NewSocket7E(socket net.Conn) *Socket7E {
	return &Socket7E{
		socket: socket,
	}
}

// Recibe la cantidad especificada de bytes del socket
func (s *Socket7E) recv(size int) ([]byte, error) {
	return safe_socket.RecvAll(s.socket, size)
}

// Envía todos los bytes especificados al socket
func (s *Socket7E) send(bytes []byte) (int, error) {
	if err := safe_socket.SendAll(s.socket, bytes); err != nil {
		return 0, err
	}

	return len(bytes), nil
}

func (s *Socket7E) sendHeader(bytes []byte) (int, error) {
	if _, err := s.send(MAGIC_NUMBER); err != nil {
		return 0, err
	}

	return s.send(bytes)
}

func (s *Socket7E) sendPayload(bytes []byte) (int, error) {
	return s.send(bytes)
}

func (s *Socket7E) ReadUntilHeaderFound() ([]byte, error) {
	for {
		b, err := s.recv(1)
		if err != nil {
			return nil, err
		}

		if b[0] == MAGIC_NUMBER[0] {
			break
		}
	}

	headerSizeBytes, err := s.recv(1)
	if err != nil {
		return nil, err
	}

	headerFields, err := s.recv(int(headerSizeBytes[0]) + 1) // +1 para el agency_id
	if err != nil {
		return nil, err
	}

	return headerFields, nil
}

func (s *Socket7E) ReadPayload(size int) ([]byte, error) {
	return s.recv(size)
}

func (s *Socket7E) Send(header []byte, payload []byte) (int, error) {
	h, err := s.sendHeader(header)
	if err != nil {
		return 0, err
	}

	p, err := s.sendPayload(payload)
	if err != nil {
		return h, err
	}

	return h + p, nil
}

// Cierra la conexión TCP con el cliente
func (s *Socket7E) Close() error {
	return s.socket.Close()
}

File: services/server/src/internal/transport/socket7e.py
import socket

from services.server.src import safe_socket


class Socket7E:

    MAGIC_NUMBER = b'\x7E'

    def __init__(self, socket: socket.socket):
        self.__socket = socket

    ''' Recibe la cantidad especificada de bytes del socket'''
    def __recv(self, size: int) -> bytes:
        return safe_socket.recv_all(self.__socket, size)

    ''' Envía todos los bytes especificados al socket'''
    def __send(self, bytes: bytes) -> int:
        return safe_socket.send_all(self.__socket, bytes)

    def __send_header(self, bytes: bytes) -> int:
        self.__send(self.MAGIC_NUMBER)
        return self.__send(bytes)

    def __send_payload(self, bytes: bytes) -> int:
        return self.__send(bytes)

    def read_until_header_found(self) -> bytes:
        while True:
            byte = self.__recv(1)
            if byte == self.MAGIC_NUMBER:
                break
        header_size_bytes = self.__recv(1)
        header_fields = self.__recv(header_size_bytes[0] + 1)  # +1 para el agency_id
        return header_fields

    def read_payload(self, size: int) -> bytes:
        return self.__recv(size)

    def send(self, header: bytes, payload: bytes) -> int:
        h = self.__send_header(header)
        p = self.__send_payload(payload)
        return h + p

    ''' Cierra la conexión TCP con el cliente'''
    def close(self) -> None:
        self.__socket.close()